File: frontend/src/lib/api/portal.ts
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { frappe } from "./frappe-client";

export function useMyProfile() {
  return useQuery({
    queryKey: ["my-profile"],
    queryFn: () => frappe.get<Record<string, unknown> & { name: string; patient_name: string }>("healthcare_erp.api.portal.my_profile"),
  });
}

export interface MyAppointment {
  name: string;
  practitioner: string;
  department: string | null;
  appointment_date: string;
  appointment_time: string;
  status: string;
  appointment_type: string | null;
}

export function useMyAppointments() {
  return useQuery({
    queryKey: ["my-appointments"],
    queryFn: () => frappe.get<MyAppointment[]>("healthcare_erp.api.portal.my_appointments"),
  });
}

export function useBookAppointment() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (vars: { practitioner: string; appointment_date: string; appointment_time: string; appointment_type?: string }) =>
      frappe.post("healthcare_erp.api.portal.book_appointment", vars),
    onSuccess: () => qc.invalidateQueries({ queryKey: ["my-appointments"] }),
  });
}

export function useCancelMyAppointment() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (name: string) => frappe.post("healthcare_erp.api.portal.cancel_my_appointment", { name }),
    onSuccess: () => qc.invalidateQueries({ queryKey: ["my-appointments"] }),
  });
}

export function useMyMedicalRecords() {
  return useQuery({
    queryKey: ["my-medical-records"],
    queryFn: () => frappe.get<Array<{ name: string; encounter_date: string; practitioner: string; chief_complaint?: string; treatment_plan?: string; follow_up_date?: string }>>(
      "healthcare_erp.api.portal.my_medical_records"
    ),
  });
}

export function useMyPrescriptions() {
  return useQuery({
    queryKey: ["my-prescriptions"],
    queryFn: () => frappe.get<Array<{ name: string; drug_name: string; dosage: string; frequency: string; refills_allowed: number; refills_used: number; status: string }>>(
      "healthcare_erp.api.portal.my_prescriptions"
    ),
  });
}

export function useMyLabResults() {
  return useQuery({
    queryKey: ["my-lab-results"],
    queryFn: () => frappe.get<Array<{ name: string; template: string; result_value: string; normal_range: string; result_date: string }>>(
      "healthcare_erp.api.portal.my_lab_results"
    ),
  });
}

export function useMyRadiologyReports() {
  return useQuery({
    queryKey: ["my-radiology-reports"],
    queryFn: () => frappe.get<Array<{ name: string; radiology_order: string; findings: string; impression: string; reported_on: string }>>(
      "healthcare_erp.api.portal.my_radiology_reports"
    ),
  });
}

export function useMyInvoices() {
  return useQuery({
    queryKey: ["my-invoices"],
    queryFn: () => frappe.get<Array<{ name: string; posting_date: string; grand_total: number; outstanding_amount: number; status: string }>>(
      "healthcare_erp.api.portal.my_invoices"
    ),
  });
}

export function useMyDocuments() {
  return useQuery({
    queryKey: ["my-documents"],
    queryFn: () => frappe.get<Array<{ name: string; file_name: string; file_url: string; file_size: number; creation: string }>>(
      "healthcare_erp.api.portal.my_documents"
    ),
  });
}

export function useUploadMyDocument(patientId: string | undefined) {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (file: File) => {
      if (!patientId) throw new Error("Patient profile not loaded yet");
      return frappe.uploadFile(file, { doctype: "Patient", docname: patientId });
    },
    onSuccess: () => qc.invalidateQueries({ queryKey: ["my-documents"] }),
  });
}

File: apps/healthcare_erp/healthcare_erp/api/portal.py
"""Module 14 — Patient Portal.

Every endpoint here resolves the *calling* user's own Patient record and
never accepts a `patient` argument from the client — the portal is a
strictly self-service surface. Record-level access still also flows
through the `Patient`-role User Permission (doc_events/patient.py), so
this is defense in depth, not the only gate.
"""

from __future__ import annotations

import frappe

from healthcare_erp.api.utils import envelope


def _current_patient() -> str:
	patient = frappe.db.get_value("Patient", {"user_id": frappe.session.user}, "name")
	if not patient:
		frappe.throw(frappe._("No patient record is linked to your account"), frappe.PermissionError)
	return patient


@frappe.whitelist()
def my_profile():
	patient = frappe.get_doc("Patient", _current_patient())
	return envelope(patient.as_dict())


@frappe.whitelist()
def my_appointments():
	patient = _current_patient()
	return envelope(frappe.get_all(
		"Patient Appointment", filters={"patient": patient},
		fields=["name", "practitioner", "department", "appointment_date", "appointment_time", "status", "appointment_type"],
		order_by="appointment_date desc", limit_page_length=100,
	))


@frappe.whitelist(methods=["POST"])
def book_appointment(practitioner: str, appointment_date: str, appointment_time: str, appointment_type: str | None = None):
	patient = _current_patient()
	appt = frappe.new_doc("Patient Appointment")
	appt.update({
		"patient": patient, "practitioner": practitioner, "appointment_date": appointment_date,
		"appointment_time": appointment_time, "appointment_type": appointment_type,
	})
	appt.insert()
	return envelope(appt.as_dict(), {"message": "Appointment booked"})


@frappe.whitelist(methods=["POST"])
def cancel_my_appointment(name: str):
	patient = _current_patient()
	doc = frappe.get_doc("Patient Appointment", name)
	if doc.patient != patient:
		frappe.throw(frappe._("Not your appointment"), frappe.PermissionError)
	doc.status = "Cancelled"
	doc.save()
	return envelope(doc.as_dict())


@frappe.whitelist()
def my_medical_records():
	patient = _current_patient()
	if not frappe.db.exists("DocType", "Patient Encounter"):
		return envelope([])
	meta = frappe.get_meta("Patient Encounter")
	fields = ["name", "encounter_date", "practitioner", "docstatus"]
	if meta.get_field("chief_complaint"):
		fields += ["chief_complaint", "treatment_plan", "follow_up_date"]
	return envelope(frappe.get_all(
		"Patient Encounter", filters={"patient": patient, "docstatus": 1}, fields=fields,
		order_by="encounter_date desc", limit_page_length=100,
	))


@frappe.whitelist()
def my_prescriptions():
	patient = _current_patient()
	if not frappe.db.exists("DocType", "Prescription Refill"):
		return envelope([])
	return envelope(frappe.get_all(
		"Prescription Refill", filters={"patient": patient},
		fields=["name", "drug_name", "dosage", "frequency", "refills_allowed", "refills_used", "status"],
		order_by="creation desc", limit_page_length=100,
	))


@frappe.whitelist()
def my_lab_results():
	patient = _current_patient()
	if not frappe.db.exists("DocType", "Lab Test"):
		return envelope([])
	return envelope(frappe.get_all(
		"Lab Test", filters={"patient": patient, "status": "Approved"},
		fields=["name", "template", "result_value", "normal_range", "result_date"],
		order_by="result_date desc", limit_page_length=100,
	))


@frappe.whitelist()
def my_radiology_reports():
	patient = _current_patient()
	if not frappe.db.exists("DocType", "Radiology Report"):
		return envelope([])
	return envelope(frappe.get_all(
		"Radiology Report", filters={"patient": patient, "status": "Final"},
		fields=["name", "radiology_order", "findings", "impression", "reported_on"],
		order_by="reported_on desc", limit_page_length=100,
	))


@frappe.whitelist()
def my_invoices():
	patient = _current_patient()
	if not frappe.db.exists("DocType", "Sales Invoice"):
		return envelope([])
	meta = frappe.get_meta("Sales Invoice")
	filters = {"patient": patient} if meta.get_field("patient") else {"customer": patient}
	return envelope(frappe.get_all(
		"Sales Invoice", filters=filters,
		fields=["name", "posting_date", "grand_total", "outstanding_amount", "status"],
		order_by="posting_date desc", limit_page_length=100,
	))


@frappe.whitelist()
def my_documents():
	patient = _current_patient()
	return envelope(frappe.get_all(
		"File", filters={"attached_to_doctype": "Patient", "attached_to_name": patient},
		fields=["name", "file_name", "file_url", "file_size", "creation"],
		order_by="creation desc", limit_page_length=100,
	))
